device/cuda/src/finding/finding_algorithm.cu

```diff
@@ -344,71 +345,74 @@ finding_algorithm<stepper_t, navigator_t>::operator()(
-        {std::vector<std::size_t>(n_tips_total,
-                                  m_cfg.max_track_candidates_per_track),
-         m_mr.main, m_mr.host, vecmem::data::buffer_type::resizable}};
-
-    m_copy.setup(track_candidates_buffer.headers)->ignore();
-    m_copy.setup(track_candidates_buffer.items)->ignore();
-
-    // Create buffer for valid indices
-    vecmem::data::vector_buffer<unsigned int> valid_indices_buffer(n_tips_total,
+    vecmem::data::vector_buffer<unsigned int> valid_tip_idx_buffer(n_tips_total,
                                                                    m_mr.main);
+    vecmem::data::vector_buffer<unsigned int> valid_tip_length_buffer(
+        n_tips_total, m_mr.main);
+    vecmem::unique_alloc_ptr<unsigned int> n_valid_tips_device =
+        vecmem::make_unique_alloc<unsigned int>(m_mr.main);
 
-    unsigned int n_valid_tracks = 0;
+    TRACCC_CUDA_ERROR_CHECK(cudaMemsetAsync(n_valid_tips_device.get(), 0,
+                                            sizeof(unsigned int), stream));
 
-    // @Note: nBlocks can be zero in case there is no tip. This happens when
-    // chi2_max config is set tightly and no tips are found
-    if (n_tips_total > 0) {
-        vecmem::unique_alloc_ptr<unsigned int> n_valid_tracks_device =
-            vecmem::make_unique_alloc<unsigned int>(m_mr.main);
-        TRACCC_CUDA_ERROR_CHECK(cudaMemsetAsync(n_valid_tracks_device.get(), 0,
-                                                sizeof(unsigned int), stream));
+    std::vector<unsigned int> valid_tips_length_host;
 
+    if (n_tips_total > 0) {
         const unsigned int nThreads = m_warp_size * 2;
         const unsigned int nBlocks = (n_tips_total + nThreads - 1) / nThreads;
 
-        kernels::build_tracks<<<nBlocks, nThreads, 0, stream>>>(
-            m_cfg, {measurements, seeds_buffer, links_buffer, tips_buffer,
-                    track_candidates_buffer, valid_indices_buffer,
-                    n_valid_tracks_device.get()});
+        kernels::count_tracks<<<nBlocks, nThreads, 0, stream>>>(
+            m_cfg,
+            {measurements, links_buffer, tips_buffer, valid_tip_idx_buffer,
+             valid_tip_length_buffer, n_valid_tips_device.get()});
+
         TRACCC_CUDA_ERROR_CHECK(cudaGetLastError());
 
-        // Global counter object: Device -> Host
+        unsigned int n_valid_tips = 0;
+
         TRACCC_CUDA_ERROR_CHECK(cudaMemcpyAsync(
-            &n_valid_tracks, n_valid_tracks_device.get(), sizeof(unsigned int),
+            &n_valid_tips, n_valid_tips_device.get(), sizeof(unsigned int),
             cudaMemcpyDeviceToHost, stream));
 
         m_stream.synchronize();
+
+        valid_tips_length_host.resize(n_valid_tips);
+
+        TRACCC_CUDA_ERROR_CHECK(cudaMemcpyAsync(
+            valid_tips_length_host.data(), valid_tip_length_buffer.ptr(),
+            n_valid_tips * sizeof(unsigned int), cudaMemcpyDeviceToHost,
+            stream));
+
+        m_stream.synchronize();
```

Comment:
Since you only need one of the vectors on the host, maybe the buffer with `tuple<uint, uint>` elements is not the best after all. But we could still make an SoA for this purpose. 🤔 In which case the low level code would still be done by VecMem, which the algorithmic code just "pushing back" into the container.